### DISCUSSION TITLE
Joining a view with another table via PostgREST/Supabase SDK

### DISCUSSION THREAD

**Discussion by newageoflight** [supabase/supabase#20028] [Questions] [answered]
As per the title. In my SQL database I have a set of tables arranged using a class table inheritance hierarchy, with several inheriting from a base class called `protoevent`. To make it easier to work with these in Supabase I've been using SQL views e.g. the view for `events` that includes the properties from the base class is called `event_list`.

The problem arises when trying to join it with the `tags` table through an intermediary join table called `protoevents_tags` which contains a list of edges between `protoevents` and `tags`. The query syntax below, for instance, fails:

```typescript
const { data, error } = await supabase
    .from("event_list")
    .select("*, tags (label)")
    .eq("id", params.eid)
    .single();
```

The error returned is:

```
{
  code: 'PGRST200',
  details: "Searched for a foreign key relationship between 'event_list' and 'tags' in the schema 'public', but no matches were found.",
  hint: "Perhaps you meant 'task_list' instead of 'tags'.",
  message: "Could not find a relationship between 'event_list' and 'tags' in the schema cache"
}
```

Within Postgres itself, however, it _is_ possible to join a view to a table. So now the question I have is:

- Why isn't this working? Is it a problem with the syntax I'm using, my underlying setup, or a limitation of Supabase/PostgREST itself?

The solution I've thought of is to reword the view query such that `protoevents` is the first operand in the `INNER JOIN` clause where the view is created. However this feels very inelegant to do given that I'd have to rely on the existence of a column that specifies the child class type of the protoevent record.

**Comment by GaryAustin1** [comment]
With a join table both foreign keys have to be part of a composite primary key.  This is nothing to do with views if that is the issue.

**Reply by newageoflight** [reply]
Thanks for your answer, but this doesn't address the problem I have. The join table `protoevents_tags` already has this composite primary key set up, as is evident from the below SQL code:

```sql
CREATE TABLE IF NOT EXISTS public.protoevents_tags (
  peid INT NOT NULL REFERENCES public.protoevents (id),
  tag_id INT NOT NULL REFERENCES public.tags (id),
  PRIMARY KEY (peid, tag_id)
);
```

On the other hand, the view is set up as follows:

```sql
CREATE OR REPLACE VIEW event_list AS
SELECT
  pe.pt_id, ... -- Other protoevent attributes that are not the id, as this is referenced in e.id
  e.*
FROM public.events e
INNER JOIN public.protoevents pe ON e.id = pe.id;
```

**Comment by newageoflight** [comment, marked as answer]
To answer my own question, this was indeed fixed by rearranging the view creation statement slightly so that the parent table came first:

```sql
CREATE OR REPLACE VIEW event_list AS
SELECT
  pe.*,
  e.start_dt, ... -- Subclass-specific attributes
FROM public.protoevents pe
INNER JOIN public.events e ON e.id = pe.id;
```
